src/main/io/statusindicator.c

```diff
@@ -20,71 +20,106 @@
 
 #include <platform.h>
 
+#include "common/axis.h"
+#include "common/maths.h"
+
+#include "config/parameter_group.h"
+#include "config/runtime_config.h"
+
 #include "drivers/system.h"
 #include "drivers/gpio.h"
 #include "drivers/light_led.h"
 #include "drivers/sound_beeper.h"
+#include "drivers/serial.h"
+#include "drivers/sensor.h"
+#include "drivers/pwm_mapping.h"
+#include "drivers/accgyro.h"
 
-#include "statusindicator.h"
-
-static uint32_t warningLedTimer = 0;
+#include "io/serial_cli.h"
+#include "io/rc_controls.h"
 
-typedef enum {
-    WARNING_LED_OFF = 0,
-    WARNING_LED_ON,
-    WARNING_LED_FLASH
-} warningLedState_e;
+#include "sensors/sensors.h"
+#include "sensors/acceleration.h"
 
-static warningLedState_e warningLedState = WARNING_LED_OFF;
+#include "flight/failsafe.h"
+#include "flight/imu.h"
 
-void warningLedResetTimer(void)
-{
-    uint32_t now = millis();
-    warningLedTimer = now + 500000;
-}
+#include "mw.h"
+#include "scheduler.h"
 
-void warningLedEnable(void)
-{
-    warningLedState = WARNING_LED_ON;
-}
+#include "statusindicator.h"
 
-void warningLedDisable(void)
-{
-    warningLedState = WARNING_LED_OFF;
-}
+static uint32_t warningLedTimer = 0;
 
-void warningLedFlash(void)
+typedef enum {
+    ARM_PREV_NONE = 0,
+    ARM_PREV_CLI,       // 2 flashes - CLI active in the configurator
+    ARM_PREV_FAILSAFE,  // 3 flashes - Failsafe mode
+    ARM_PREV_ANGLE,     // 4 flashes - Maximum arming angle exceeded
+    ARM_PREV_CALIB,     // 5 flashes - Calibration active
+    ARM_PREV_OVERLOAD   // 6 flashes - System overload
+} armingPreventedReason_e;
+
+static armingPreventedReason_e armingPreventionReason = ARM_PREV_NONE;
+static uint8_t flashesLeft = 0;
+
+armingPreventedReason_e getArmingPreventionReason(void)
 {
-    warningLedState = WARNING_LED_FLASH;
+    if (isCalibrating()) {
+        return ARM_PREV_CALIB;
+    }
+    if (rcModeIsActive(BOXFAILSAFE) || failsafePhase() == FAILSAFE_LANDED) {
+        return ARM_PREV_FAILSAFE;
+    }
+    if (!imuIsAircraftArmable(armingConfig()->max_arm_angle)) {
+        return ARM_PREV_ANGLE;
+    }
+    if (cliMode) {
+        return ARM_PREV_CLI;
+    }
+    if (isSystemOverloaded()) {
+        return ARM_PREV_OVERLOAD;
+    }
+    return ARM_PREV_NONE;
 }
 
 void warningLedRefresh(void)
 {
-    switch (warningLedState) {
-        case WARNING_LED_OFF:
+    if (flashesLeft == 0) {
+        armingPreventionReason = getArmingPreventionReason();
```

Comment:
you can set `flashesLeft` (persistent state) here. It will simplify code structure.

(yours is:
```
if(A)
   B = state();
if(B)
  if(A)
     A = fn(B);
```
)